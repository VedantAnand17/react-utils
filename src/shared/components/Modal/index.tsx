/* global document */

import { noop } from 'lodash';

import {
  type ReactNode,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';

import ReactDom from 'react-dom';
import PT from 'prop-types';
import themed, { type Theme } from '@dr.pogodin/react-themes';

import baseTheme from './base-theme.scss';
import S from './styles.scss';

const validThemeKeys = ['container', 'overlay'] as const;

type PropsT = {
  cancelOnScrolling?: boolean;
  children?: ReactNode;
  containerStyle?: React.CSSProperties;
  dontDisableScrolling?: boolean;
  onCancel?: () => void;
  theme: Theme<typeof validThemeKeys>;
};

/**
 * The `<Modal>` component implements a simple themeable modal window, wrapped
 * into the default theme. `<BaseModal>` exposes the base non-themed component.
 * **Children:** Component children are rendered as the modal content.
 * @param {object} props Component properties. Beside props documented below,
 *  [Other theming properties](https://www.npmjs.com/package/@dr.pogodin/react-themes#themed-component-properties) are supported as well.
 * @param {function} [props.onCancel] The callback to trigger when user
 * clicks outside the modal, or presses Escape. It is expected to hide the
 * modal.
 * @param {ModalTheme} [props.theme] _Ad hoc_ theme.
 */
const BaseModal: React.FunctionComponent<PropsT> = ({
  cancelOnScrolling,
  children,
  containerStyle,
  dontDisableScrolling,
  onCancel,
  theme,
}) => {
  const containerRef = useRef<HTMLDivElement | null>(null);
  const overlayRef = useRef<HTMLDivElement | null>(null);
  const [portal, setPortal] = useState<HTMLDivElement>();

  useEffect(() => {
    const p = document.createElement('div');
    document.body.appendChild(p);
    setPortal(p);
    return () => {
      document.body.removeChild(p);
    };
  }, []);

  // Sets up modal cancellation of scrolling, if opted-in.
  useEffect(() => {
    if (cancelOnScrolling && onCancel) {
      window.addEventListener('scroll', onCancel);
    }
    return () => {
      if (cancelOnScrolling && onCancel) {
        window.removeEventListener('scroll', onCancel);
      }
    };
  }, [cancelOnScrolling, onCancel]);

  // Disables window scrolling, if it is not opted-out.
  useEffect(() => {
    if (!dontDisableScrolling) {
      document.body.classList.add(S.scrollingDisabledByModal);
    }
    return () => {
      if (!dontDisableScrolling) {
        document.body.classList.remove(S.scrollingDisabledByModal);
      }
    };
  }, [dontDisableScrolling]);

  const focusLast = useMemo(() => (
    <div
      onFocus={() => {
        const elems = containerRef.current?.querySelectorAll('*') as NodeListOf<HTMLElement>;
        for (let i = elems.length - 1; i >= 0; --i) {
          elems[i].focus();
          if (document.activeElement === elems[i]) return;
        }
        overlayRef.current?.focus();
      }}
      /* eslint-disable jsx-a11y/no-noninteractive-tabindex */
      tabIndex={0}
      /* eslint-enable jsx-a11y/no-noninteractive-tabindex */
    />
  ), []);

  return portal ? ReactDom.createPortal(
    (
      <>
        {focusLast}
        <div
          aria-label="Cancel"
          className={theme.overlay}
          onClick={() => onCancel && onCancel()}
          onKeyDown={(e) => {
            if (e.key === 'Escape' && onCancel) onCancel();
          }}
          ref={(node) => {
            if (node && node !== overlayRef.current) {
              overlayRef.current = node;
              node.focus();
            }
          }}
          role="button"
          tabIndex={0}
        />
        <div
          aria-modal="true"
          className={theme.container}
          onWheel={(event) => event.stopPropagation()}
          ref={containerRef}
          role="dialog"
          style={containerStyle}
        >
          {children}
        </div>
        <div
          onFocus={() => {
            overlayRef.current?.focus();
          }}
          /* eslint-disable jsx-a11y/no-noninteractive-tabindex */
          tabIndex={0}
          /* eslint-enable jsx-a11y/no-noninteractive-tabindex */
        />
        {focusLast}
      </>
    ),
    portal,
  ) : null;
};

const ThemedModal = themed(
  BaseModal,
  'Modal',
  validThemeKeys,
  baseTheme,
);

BaseModal.propTypes = {
  cancelOnScrolling: PT.bool,
  children: PT.node,
  containerStyle: PT.shape({}),
  dontDisableScrolling: PT.bool,
  onCancel: PT.func,
  theme: ThemedModal.themeType.isRequired,
};

BaseModal.defaultProps = {
  cancelOnScrolling: false,
  children: null,
  containerStyle: undefined,
  dontDisableScrolling: false,
  onCancel: noop,
};

export default ThemedModal;

/* Non-themed version of the Modal. */
export { BaseModal };
