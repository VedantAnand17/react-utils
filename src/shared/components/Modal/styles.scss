:global {
  body.scrolling-disabled-by-modal {
    overflow: hidden;
  }
}
