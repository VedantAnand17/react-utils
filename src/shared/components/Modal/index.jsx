/**
 * This generic component will implement the semi-transparent background
 * and the white window in the center, which wraps the content provided as
 * children.
 *
 * When semi-transparent background is clicked, it should trigger the onCancel()
 * callback passed from the parent.
 */

/* global document */

import _ from 'lodash';
import React, { useMemo } from 'react';
import ReactDom from 'react-dom';
import PT from 'prop-types';
import themed from '@dr.pogodin/react-themes';

import baseTheme from './base-theme.scss';
import './styles.scss';

function BaseModal({
  children,
  onCancel,
  theme,
}) {
  const containerRef = React.useRef();
  const overlayRef = React.useRef();
  const [portal, setPortal] = React.useState();

  React.useEffect(() => {
    const p = document.createElement('div');
    document.body.classList.add('scrolling-disabled-by-modal');
    document.body.appendChild(p);
    setPortal(p);
    return () => {
      document.body.classList.remove('scrolling-disabled-by-modal');
      document.body.removeChild(p);
    };
  }, []);

  const focusLast = useMemo(() => (
    <div
      onFocus={() => {
        const elems = containerRef.current.querySelectorAll('*');
        for (let i = elems.length - 1; i >= 0; --i) {
          elems[i].focus();
          if (document.activeElement === elems[i]) return;
        }
        overlayRef.current.focus();
      }}
      /* eslint-disable jsx-a11y/no-noninteractive-tabindex */
      tabIndex="0"
      /* eslint-enable jsx-a11y/no-noninteractive-tabindex */
    />
  ), []);

  return portal ? ReactDom.createPortal(
    (
      <>
        {focusLast}
        <div
          aria-label="Cancel"
          className={theme.overlay}
          onClick={() => onCancel()}
          onKeyDown={(e) => {
            if (e.key === 'Escape') onCancel();
          }}
          ref={(node) => {
            if (node && node !== overlayRef.current) {
              overlayRef.current = node;
              node.focus();
            }
          }}
          role="button"
          tabIndex="0"
        />
        <div
          aria-modal="true"
          className={theme.container}
          onWheel={(event) => event.stopPropagation()}
          ref={containerRef}
          role="dialog"
        >
          {children}
        </div>
        <div
          onFocus={() => {
            overlayRef.current.focus();
          }}
          /* eslint-disable jsx-a11y/no-noninteractive-tabindex */
          tabIndex="0"
          /* eslint-enable jsx-a11y/no-noninteractive-tabindex */
        />
        {focusLast}
      </>
    ),
    portal,
  ) : null;
}

const ThemedModal = themed(
  'Modal',
  [
    'container',
    'overlay',
  ],
  baseTheme,
)(BaseModal);

BaseModal.propTypes = {
  onCancel: PT.func,
  children: PT.node,
  theme: ThemedModal.themeType,
};

BaseModal.defaultProps = {
  onCancel: _.noop,
  children: null,
};

export default ThemedModal;

/* Non-themed version of the Modal. */
export { BaseModal };
