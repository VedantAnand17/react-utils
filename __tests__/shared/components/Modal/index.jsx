import React from 'react';

import pretty from 'pretty';

import Modal from 'components/Modal';

import {
  act,
  mount,
  unmount,
  simulate,
} from 'utils/jest';

jest.useFakeTimers();

let scene;
let onCancel;
beforeEach(() => {
  onCancel = jest.fn();
  scene = mount((
    <div data-id="Modal Code Parent">
      <div data-id="Modal Code Sibling" />
      <Modal
        onCancel={onCancel}
        theme={{
          ad: 'ad',
          hoc: 'hoc',
          context: 'context',
          container: 'container',
        }}
      >
        Modal Content
      </Modal>
    </div>
  ));
});

afterEach(() => {
  if (scene) {
    unmount(scene);
    scene = null;
  }
});

test('Snapshot match', () => {
  expect(pretty(document.body.innerHTML)).toMatchSnapshot();
  act(() => {
    jest.runAllTimers();
  });
  expect(pretty(document.body.innerHTML)).toMatchSnapshot();
});

test('onCancel', () => {
  act(() => {
    jest.runAllTimers();
  });
  const overlay = document.querySelector('div[aria-label=Cancel]');
  simulate.click(overlay);
  expect(onCancel).toHaveBeenCalled();
});

test('onWheel', () => {
  act(() => {
    jest.runAllTimers();
  });
  const container = document.getElementsByClassName('container');
  expect(container.length).toBe(1);
  const stopPropagation = jest.fn();
  simulate.wheel(container[0], { stopPropagation });
  expect(stopPropagation).toHaveBeenCalled();
});
