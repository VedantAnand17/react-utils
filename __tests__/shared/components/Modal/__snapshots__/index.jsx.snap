// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot match 1`] = `
"<div>
  <div data-id=\\"Modal Code Parent\\">
    <div data-id=\\"Modal Code Sibling\\"></div>
  </div>
</div>"
`;

exports[`Snapshot match 2`] = `
"<div>
  <div data-id=\\"Modal Code Parent\\">
    <div data-id=\\"Modal Code Sibling\\"></div>
  </div>
</div>
<div>
  <div tabindex=\\"0\\"></div>
  <div aria-label=\\"Cancel\\" class=\\"-dr-pogodin-react-utils___src-shared-components-Modal-base-theme___overlay___1z_lT2\\" role=\\"button\\" tabindex=\\"0\\"></div>
  <div aria-modal=\\"true\\" class=\\"-dr-pogodin-react-utils___src-shared-components-Modal-base-theme___container___uS4uDL ad hoc container\\" role=\\"dialog\\">Modal Content</div>
  <div tabindex=\\"0\\"></div>
  <div tabindex=\\"0\\"></div>
</div>"
`;
